quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/legacy/infinispan/CacheManagerFactory.java

```diff
@@ -173,6 +181,74 @@ private RemoteCacheManager startRemoteCacheManager() {
+        var baseConfig = defaultRemoteCacheBuilder().build();
+
+        Stream.of(USER_SESSION_CACHE_NAME, OFFLINE_USER_SESSION_CACHE_NAME, CLIENT_SESSION_CACHE_NAME, OFFLINE_CLIENT_SESSION_CACHE_NAME, ACTION_TOKEN_CACHE, WORK_CACHE_NAME)
+                .forEach(name -> builder.remoteCache(name).configuration(baseConfig.toStringConfiguration(name)));
+
+        // We need indexed caches because the delete statement fails for non-indexed cache.
+        createIndexedRemoteCache(builder, LOGIN_FAILURE_CACHE_NAME, List.of(LoginFailureEntity.class));
+        createIndexedRemoteCache(builder, AUTHENTICATION_SESSIONS_CACHE_NAME, List.of(RootAuthenticationSessionEntity.class));
+    }
+
+    private static void createIndexedRemoteCache(org.infinispan.client.hotrod.configuration.ConfigurationBuilder builder, String name, List<Class<?>> entities) {
+        var config = indexedRemoteCacheBuilder(entities).build();
+        builder.remoteCache(name).configuration(config.toStringConfiguration(name));
+    }
+
+    private static ConfigurationBuilder defaultRemoteCacheBuilder() {
+        var builder = new ConfigurationBuilder();
+        builder.clustering().cacheMode(CacheMode.DIST_SYNC);
+        builder.encoding().mediaType(MediaType.APPLICATION_PROTOSTREAM);
+        return builder;
+    }
+
+    private static ConfigurationBuilder indexedRemoteCacheBuilder(List<Class<?>> entities) {
+        var builder = defaultRemoteCacheBuilder();
+        var indexBuilder = builder.indexing().enable();
+        entities.stream()
+                .map(Marshalling::protoEntity)
+                .forEach(indexBuilder::addIndexedEntity);
+        return builder;
+    }
+
+    private void updateProtoSchema(RemoteCacheManager remoteCacheManager) {
+        var key = KeycloakModelSchema.INSTANCE.getProtoFileName();
+        var current = KeycloakModelSchema.INSTANCE.getProtoFile();
+
+        RemoteCache<String, String> protostreamMetadataCache = remoteCacheManager.getCache(InternalCacheNames.PROTOBUF_METADATA_CACHE_NAME);
+        var stored = protostreamMetadataCache.getWithMetadata(key);
+        if (stored == null) {
+            if (protostreamMetadataCache.putIfAbsent(key, current) == null) {
+                logger.info("Infinispan Protostream schema uploaded for the first time.");
+            } else {
+                logger.info("Failed to update Infinispan Protostream schema. Assumed it was updated by other Keycloak server.");
+            }
+            return;
+        }
+        if (Objects.equals(stored.getValue(), current)) {
+            logger.info("Infinispan Protostream schema is up to date!");
+            return;
+        }
+        if (protostreamMetadataCache.replaceWithVersion(key, current, stored.getVersion())) {
+            logger.info("Infinispan Protostream schema successful updated.");
+            reindexCaches(remoteCacheManager);
```

Comment:
re-indexing is blocking and it will block the KC startup (on the first node). I think Infinispan can handle reads/writes while reindexing, but I'm not sure if queries return the correct value. 

Re-index is only required when the indexing schema changes. KC will trigger that and no manual intervention should be necessary.

I'll let @fax4ever comment on the Infinispan future.

Reply:
As you write "will block the KC startup on the first node, and only happens on a schema change", this resolves it for me. Thank you for adding these details.

Reply:
I confirm what @pruivo said